File: examples/langchain_agent.py
"""
LangChain ReAct Agent — fully traced by OrchestraAI.

Auto-captures: LLM calls (with tokens), tool invocations, agent reasoning,
retriever calls, chain start/end — all as nested spans.

Prerequisites:
    pip install -e sdks/python
    pip install langchain-openai langchain-core

Run:
    export ORCHESTRA_API_KEY=oai_...
    export OPENAI_API_KEY=sk-...
    python examples/langchain_agent.py
"""

import os
from orchestra_ai import OrchestraAI
from orchestra_ai.integrations.langchain_tracer import get_handler

from langchain_openai import ChatOpenAI
from langchain_core.tools import tool
from langchain_core.messages import HumanMessage

# ── Setup ──────────────────────────────────────────────────────
oa = OrchestraAI(
    api_key=os.environ["ORCHESTRA_API_KEY"],
    base_url=os.getenv("ORCHESTRA_BASE_URL", "http://localhost:3001"),
)

# Create a handler — all LangChain callbacks are forwarded to OrchestraAI
handler = get_handler(oa, agent_name="langchain-research-agent", session_id="demo-session")


# ── Tools ──────────────────────────────────────────────────────
@tool
def get_weather(city: str) -> str:
    """Get the current weather for a city."""
    data = {"san francisco": "Foggy, 58F", "new york": "Sunny, 72F", "london": "Rainy, 55F"}
    return data.get(city.lower(), f"No data for {city}")


@tool
def calculator(expression: str) -> str:
    """Evaluate a math expression safely."""
    allowed = set("0123456789+-*/.(). ")
    if not all(c in allowed for c in expression):
        return "Invalid expression"
    return str(eval(expression))


# ── Run ────────────────────────────────────────────────────────
llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)
llm_with_tools = llm.bind_tools([get_weather, calculator])

print("=== LangChain Agent with OrchestraAI Tracing ===\n")
print("Sending query: 'What's the weather in SF? Also what is 42 * 17?'\n")

response = llm_with_tools.invoke(
    [HumanMessage(content="What's the weather in San Francisco? Also what is 42 * 17?")],
    config={"callbacks": [handler]},
)

print(f"Response: {response.content or '(tool calls pending)'}")
if response.tool_calls:
    print(f"Tool calls: {[tc['name'] for tc in response.tool_calls]}")

print("\nCheck traces at http://localhost:3000/dashboard/traces")
oa.close()

File: examples/langchain_agent.ts
/**
 * LangChain Agent — traced by OrchestraAI (TypeScript).
 *
 * Auto-captures: LLM calls, tool invocations, chain start/end, retrievers.
 *
 * Prerequisites:
 *   npm link ./sdks/typescript
 *   npm install @langchain/openai @langchain/core
 *
 * Run:
 *   export ORCHESTRA_API_KEY=oai_...
 *   export OPENAI_API_KEY=sk-...
 *   npx tsx examples/langchain_agent.ts
 */

import { OrchestraAI } from '../sdks/typescript/src';
import { createLangChainHandler } from '../sdks/typescript/src/integrations/langchain';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage } from '@langchain/core/messages';
import { tool } from '@langchain/core/tools';
import { z } from 'zod';

const oa = new OrchestraAI({
  apiKey: process.env.ORCHESTRA_API_KEY!,
  baseUrl: process.env.ORCHESTRA_BASE_URL || 'http://localhost:3001',
});

// Create OrchestraAI callback handler
const handler = createLangChainHandler({
  client: oa,
  agentName: 'langchain-ts-agent',
  metadata: { example: 'langchain_agent.ts' },
});

// Define tools
const getWeather = tool(
  async ({ city }: { city: string }) => {
    const data: Record<string, string> = {
      'san francisco': 'Foggy, 58F',
      'new york': 'Sunny, 72F',
    };
    return data[city.toLowerCase()] || `No data for ${city}`;
  },
  {
    name: 'get_weather',
    description: 'Get the current weather for a city',
    schema: z.object({ city: z.string() }),
  }
);

async function main() {
  console.log('=== LangChain Agent with OrchestraAI Tracing (TypeScript) ===\n');

  const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });
  const llmWithTools = llm.bindTools([getWeather]);

  const response = await llmWithTools.invoke(
    [new HumanMessage("What's the weather in San Francisco?")],
    { callbacks: [handler] },
  );

  console.log(`Response: ${response.content || '(tool calls pending)'}`);
  console.log('\nCheck traces at http://localhost:3000/dashboard/traces');
}

main().catch(console.error);
